File: kunal/1043. Partition Array for Maximum Sum.java
class Solution {
    public int maxSum(int[] dp, int[] arr, int start, int end, int k) {
        int maxValue = Integer.MIN_VALUE;
        int max = Integer.MIN_VALUE;

        while (start >= 0 && end - start + 1 <= k) {
            maxValue = Math.max(maxValue, arr[start]);

            max = Math.max(max, dp[start] + ((end - start + 1) * maxValue));

            start--;
        }

        return max;
    }

    public int maxSumAfterPartitioning(int[] arr, int k) {
        int[] dp = new int[arr.length + 1];
        dp[0] = 0;

        for (int i = 0; i < arr.length; i++) {
            // Exclude from the ongoing subarray
            int exclude = dp[i] + arr[i];

            // Include in the ongoing array
            int include = maxSum(dp, arr, i, i, k);

            dp[i + 1] = Math.max(exclude, include);
        }

        return dp[dp.length - 1];
    }
}

File: kunal/1043. Partition Array for Maximum Sum.py
class Solution:
    def maxSum(self, dp, arr, start, end, k) -> int:
        maxValue = float("-inf")
        maximumSum = float("-inf")

        while start >= 0 and (end - start + 1) <= k:
            maxValue = max(maxValue, arr[start])
            maximumSum = max(maximumSum, dp[start] + ((end - start + 1) * maxValue))
            start -= 1

        return maximumSum

    def maxSumAfterPartitioning(self, arr: list[int], k: int) -> int:
        dp = []
        dp.append(0)

        for i in range(len(arr)):
            # exclude arr[i] from the ongoing subarray
            exclude = dp[i] + arr[i]

            # include arr[i] in the ongoing subarray
            include = self.maxSum(dp, arr, i, i, k)

            dp.append(max(exclude, include))

        return dp[-1]


print(Solution().maxSumAfterPartitioning([1, 15, 7, 9, 2, 5, 10], 3))
print(Solution().maxSumAfterPartitioning([1, 4, 1, 5, 7, 3, 6, 1, 9, 9, 3], 4))
print(Solution().maxSumAfterPartitioning([1], 1))
print(Solution().maxSumAfterPartitioning([6, 1, 9, 9, 3], 4))
